server/src/main/java/org/opensearch/index/translog/transfer/TransferSnapshotProvider.java

```diff
@@ -0,0 +1,141 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.translog.transfer;
+
+import org.opensearch.common.collect.Tuple;
+import org.opensearch.index.translog.TranslogReader;
+
+import java.io.IOException;
+import java.nio.file.Path;
+import java.util.HashSet;
+import java.util.LinkedList;
+import java.util.List;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+import java.util.stream.LongStream;
+
+import static org.opensearch.index.translog.transfer.FileSnapshot.TransferFileSnapshot;
+import static org.opensearch.index.translog.transfer.FileSnapshot.TranslogFileSnapshot;
+import static org.opensearch.index.translog.transfer.FileSnapshot.CheckpointFileSnapshot;
+
+/**
+ * Provider for a {@link TransferSnapshot} which builds the snapshot from the translog and checkpoint files present on the local-disk
+ *
+ * @opensearch.internal
+ */
+public class TransferSnapshotProvider implements Supplier<TransferSnapshot> {
+
+    private final TranslogCheckpointTransferSnapshot translogTransferSnapshot;
+
+    public TransferSnapshotProvider(
+        long primaryTerm,
+        long generation,
+        Path location,
+        List<TranslogReader> readers,
+        Function<Long, String> checkpointGenFileNameMapper
+    ) throws IOException {
+        translogTransferSnapshot = new TranslogCheckpointTransferSnapshot(primaryTerm, generation, readers.size());
+        for (TranslogReader reader : readers) {
+            final long readerGeneration = reader.getGeneration();
+            final long readerPrimaryTerm = reader.getPrimaryTerm();
+            final long minTranslogGeneration = reader.getCheckpoint().getMinTranslogGeneration();
+            final long checkpointGeneration = reader.getCheckpoint().getGeneration();
+            Path translogPath = reader.path();
+            Path checkpointPath = location.resolve(checkpointGenFileNameMapper.apply(readerGeneration));
+            translogTransferSnapshot.add(
+                new TranslogFileSnapshot(readerPrimaryTerm, readerGeneration, translogPath),
+                new CheckpointFileSnapshot(readerPrimaryTerm, checkpointGeneration, minTranslogGeneration, checkpointPath)
+            );
+        }
+    }
+
+    public TranslogCheckpointTransferSnapshot get() {
+        translogTransferSnapshot.assertInvariants();
```

Comment:
I don't see this class being used anywhere so I don't have the full context, but a couple of comments (if this PR is still a work-in-progress please let me know...):

- I'd recommend using a builder pattern here, where the nested class is some kind of "TransferSnapshotBuilder" and you enforce the invariants in the build() method (by throwing an exception on failure as opposed to using the conditionally-enabled assert keyword). Then it is impossible for an invalid TransferSnapshot instance to ever exist, and you don't have to worry about doing asserts at the right time.
- Since the data is created at construction time of the supplier, why do you need a supplier at all? Usually that pattern enables deferring the work but that is not happening here.

Reply:
That's fair, will move the construction time build to the supplier and let the supplier have those enforcements

Reply:
Changed my mind, the Builder pattern I agree would work better @andrross. While I have gone ahead and retained asserts as I think this is re-verifying the consistency part. I will add tests for this in a follow up PR as a part of https://github.com/opensearch-project/OpenSearch/issues/5259

Reply:
But why use the `assert` keyword instead of unconditionally throwing an exception? I know it's quite common in this code base, but the only thing `assert` provides versus just throwing an exception is the ability to turn off the check, which seems undesirable unless you believe there is a scenario where the invariants could be violated and you would not want to fail. The only other case is if the check is expensive because I think asserts are not evaluated at runtime when disabled.

Reply:
I guess asserts generally helps validate inherent assumptions that are implictly made, which should be covered as a part of integ and unit tests. The likelihood of this happening only on production systems(where asserts are disabled) is typically low. Also in this case there might be some extra computation needed just for assertions which would otherwise be wasteful and impact latencies.